lualib/resty/kong/log.lua

```diff
@@ -0,0 +1,85 @@
+-- Copyright 2019-2022 Kong Inc.
+
+-- Licensed under the Apache License, Version 2.0 (the "License");
+-- you may not use this file except in compliance with the License.
+-- You may obtain a copy of the License at
+
+--    http://www.apache.org/licenses/LICENSE-2.0
+
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+
+
+local _M = {}
+
+
+local ffi = require("ffi")
+local base = require("resty.core.base")
+base.allows_subsystem("http")
+
+
+ffi.cdef([[
+int ngx_http_lua_kong_ffi_set_log_level(ngx_http_request_t *r, int level);
+]])
+
+
+local get_phase = ngx.get_phase
+local type = type
+local error = error
+local tostring = tostring
+local C = ffi.C
+local get_request = base.get_request
+
+
+local NGX_OK = ngx.OK
+
+
+do
+    local LEVELS = {
+        debug = ngx.DEBUG,
+        info = ngx.INFO,
+        notice = ngx.NOTICE,
+        warn = ngx.WARN,
+        err = ngx.ERR,
+        crit = ngx.CRIT,
+        alert = ngx.ALERT,
+        emerg = ngx.EMERG,
+    }
+
+    local ALLOWED_PHASES = {
+        ["balancer"] = true,
+        ["access"] = true,
+        ["content"] = true,
+    }
+
+    function _M.set_log_level(level)
+        if not ALLOWED_PHASES[get_phase()] then
+            error("API disabled in the current context", 2)
+        end
+
+        local type_level = type(level)
+        if type_level ~= "string" then
+            error("incorrect level, expects a string, got " .. type_level, 2)
+        end
+
+        local ngx_level = LEVELS[level]
+        if not ngx_level then
+            error("invalid level " .. level, 2)
+        end
+
+        local r = get_request()
```

Comment:
need to check nil or `r`

Reply:
In someplace we have no `nil` check for `get_request()`,
do we need to check and fix them?

Reply:
@chronolaw yes I think that will be a good idea.

Reply:
@chronolaw @fffonion 

https://github.com/Kong/lua-kong-nginx-module/blob/master/lualib/resty/kong/tls.lua#L74-L75

>   -- no need to check if r is nil as phase check above
>   -- already ensured it

Reply:
This is true, it won't be an issue normally; adding nil check here is more for protective purpose